Propose an AWS based IPv6 development environment

russellb opened the pull request.

abhinavdahiya commented on enhancements/aws-ipv6-dev-env.md:
```diff
@@ -0,0 +1,118 @@
+
+A PR with an implementation and documentation can be found here:
+
+* <https://github.com/openshift/installer/pull/2555>
```
IMO this doesn't seem right, personally the enhacement should drive the implementation not the other way around.
I have no way of commenting if the process is flawed because I don't have the process documented here.

markmc replied:
If you mean it would be a good idea to have more details in this doc so it can be discussed separately from the implementation ... then that makes sense

It would probably work well to take the content from the "AWS Network Environment" section of https://github.com/openshift/installer/blob/bd498521d3cd9936adaa02f700e55c57896e9067/docs/dev/aws_ipv6.md and elaborate on why that design makes sense ... and a lot of the rationale can be pulled from Russell's nice commit messages

squeed commented on enhancements/aws-ipv6-dev-env.md:
```diff
@@ -0,0 +1,158 @@
+
+Note that an IPv4 CIDR is still used for `machineCIDR` since AWS will provide a
+dual-stack (IPv4 and IPv6) environment.  We must specify the IPv4 CIDR and AWS
+will automatically allocate an IPv6 CIDR.
```
Question: will the ipv6 machine addresses be stable?

russellb replied:
Good question - yes, it looks like they are stable.  A specific address is assigned to each host, just like IPv4.  The same address should always be returned via DHCP(v6)

danwinship commented on enhancements/aws-ipv6-dev-env.md:
```diff
@@ -0,0 +1,158 @@
+OKD.
+
+We can never block IPv4 completely on AWS.  At a minimum, IPv4 is required for
+the AWS instances to reach the EC2 metadata API.
```
For metadata access, the nodes just need a link-local address, and don't need a default route. (Hm... `machineCIDR: 169.254.0.0/16` ?)

abhinavdahiya commented on enhancements/aws-ipv6-dev-env.md:
```diff
@@ -0,0 +1,158 @@
+support a dual-stack (IPv4 and IPv6) environment, so that’s what is enabled
+here.  This is a summary of the changes to the network environment:
+
+* The VPC has IPv6 enabled and a `/56` IPv6 CIDR will be allocated by AWS.
```
Maybe we can look at using bring-your-own-vpc for ipv6 networking on AWS. Because that already handles the case of user provided routes/natgw/igw etc..

danwinship replied:
Wouldn't that just be *more* work?

AFAICT you can't get a 100% single-stack IPv6 environment on AWS. eg, "[We do not provide DNS hostnames for IPv6 addresses.](https://docs.aws.amazon.com/vpc/latest/userguide/vpc-dns.html)"

But that's ok, this is helping us fix upstream for the dual-stack-but-preferring-IPv6 case, which we eventually need to support.

lihongan commented on enhancements/aws-ipv6-dev-env.md:
```diff
@@ -0,0 +1,158 @@
+  - cidr: fd01::/48
+    hostPrefix: 64
+  machineCIDR: 10.0.0.0/16
+  networkType: OVNKubernetes
```
Will not support `OpenShiftSDN` with IPv6?

danwinship replied:
That is the plan at this time

lihongan commented on enhancements/aws-ipv6-dev-env.md:
```diff
@@ -0,0 +1,158 @@
+* The IPv4 NLBs are disabled.
+* IPv6 DNS records (AAAA) are created and the IPv4 (A) records are disabled.
+* IPv6 routing is configured.  Since all instances get global IPv6 addresses,
+  NAT is not used from the instances out to the internet.
```
What about ELB for ingress (haproxy router)? will support IPv6 only or both?

russellb replied:
I've updated this doc.  The load balancer usage has changed back to IPv4 load balancers only, as AWS does not have a usable TCP load balancer with Ipv6 support.

abhinavdahiya commented on enhancements/aws-ipv6-dev-env.md:
````diff
@@ -0,0 +1,163 @@
+environment variable before running the installer:
+
+```bash
+    export OPENSHIFT_AWS_USE_IPV6=”true”
````
`OPENSHIFT_INSTALL_AWS_USE_IPV6`

russellb replied:
OK, I'll fix that in the installer PR, too.

abhinavdahiya commented on enhancements/aws-ipv6-dev-env.md:
```diff
@@ -0,0 +1,163 @@
+  IPv6, other than classic load balancers with EC2-Classic, and not EC2-VPC.
+  AWS Application Load Balancers supposedly support IPv6, but that would
+  require doing HTTPS load balancing for the API instead of just TCP load
+  balancing, so we just use the IPv4 NLBs.  API access within the cluster is
```
> API access within the cluster is still exercising IPv6.

hmm, is this over k8s networking (service IP) ?

russellb replied:
I've clarified that it's IPv6 when accessed via the Service IP.

abhinavdahiya commented on enhancements/aws-ipv6-dev-env.md:
```diff
@@ -0,0 +1,163 @@
+  require doing HTTPS load balancing for the API instead of just TCP load
+  balancing, so we just use the IPv4 NLBs.  API access within the cluster is
+  still exercising IPv6.
+* IPv6 DNS records (AAAA) are created and the IPv4 (A) records are disabled,
```
> IPv4 (A) records are disabled

what does this mean?

russellb replied:
It means we put IPv6 addresses in DNS instead of IPv4 where possible.  :-)

I'll expand on it further clarifying why the API records are still IPv4.

abhinavdahiya commented on enhancements/aws-ipv6-dev-env.md:
```diff
@@ -0,0 +1,163 @@
+  still exercising IPv6.
+* IPv6 DNS records (AAAA) are created and the IPv4 (A) records are disabled,
+  except for the API.
+* IPv6 routing is configured.  Since all instances get global IPv6 addresses,
```
from https://aws.amazon.com/premiumsupport/knowledge-center/configure-private-ipv6-subnet/

> If you want your IPv6 address to access the internet, but you don't want resources on the internet to initiate communication with your instance, you can use an egress-only internet gateway.
> -   Create an egress-only internet gateway in your Amazon VPC.
> -   Add a route to your route table that points all IPv6 traffic (::/0) or a specific range of IPv6 addresses to the egress-only internet gateway.

Can we make sure we have `resources on the internet to initiate communication with your instance` behavior required.

russellb replied:
Yeah, that's an interesting idea.  It looks like security groups are already set up to effectively provide the same result.  It seems like that's probably good enough for our purposes?

abhinavdahiya replied:
>  It looks like security groups are already set up to effectively provide the same result

personally our users will expect we follow AWS best-practices. security groups are great but they can easily _fungible_. something more fundamental that provide the _private_ behavior would be much better appreciated by our users.

> It seems like that's probably good enough for our purposes?

If we can at least look into AWS recommendations and see if they are small enough to do now that would be best.. but if we should mention it as future work in the doc at least.

russellb replied:
Well, there's no users for this except as a dev/test platform, not to be supported for any customer use.  I get your point that an egress-only gateway is much more explicit and clear about the intended behavior though.  I'll change to that.  I'll make it clear in this doc that this minute it's a todo item, but I should be able to have it done Monday if not sometime sooner this weekend.

abhinavdahiya commented:
It would be great if we can provide details around the extra CoreDNS instance? from https://github.com/openshift/installer/pull/2555/files#diff-39ee878d999c07c3e18370e612a8766dR316-R318

Also we have 9000-9990 is a ip range that is reserved for Openshift services for in-cluster communication. can we make sure we use a port from that range.

abhinavdahiya commented:
Also can we provide details on what addresses are availble to the ec2 instance, and what kubelet provides as the address to the node object.

Some details around etcd would be nice.

russellb commented:
> It would be great if we can provide details around the extra CoreDNS instance? from https://github.com/openshift/installer/pull/2555/files#diff-39ee878d999c07c3e18370e612a8766dR316-R318

I've added some more info about this.
 
> Also we have 9000-9990 is a ip range that is reserved for Openshift services for in-cluster communication. can we make sure we use a port from that range.

The way we've implemented the hack has actually changed, so the new security group rule for 5353 should no longer be needed.  All of the port 5353 traffic will be IPv6 over the SDN, and the IPv4 traffic will be IPv4 to AWS' DNS servers.

> Also can we provide details on what addresses are availble to the ec2 instance, and what kubelet provides as the address to the node object.
> 
> Some details around etcd would be nice.

Added.

abhinavdahiya commented:
One more thing, how do we know that we have a cluster in the wild that is using this dev feature.. to make sure we can filter these clusters out from our telemetry dashboards..?

russellb commented:
> One more thing, how do we know that we have a cluster in the wild that is using this dev feature.. to make sure we can filter these clusters out from our telemetry dashboards..?

That's a good question.  Right now the only thing I can think is filtering out AWS clusters that have IPv6 cluster and service CIDRs.

danwinship commented:
> One more thing, how do we know that we have a cluster in the wild that is using this dev feature.. to make sure we can filter these clusters out from our telemetry dashboards..?

We could just only allow the option to be used with `devcluster.openshift.com` (plus whatever ci uses)?

cgwalters commented on enhancements/aws-ipv6-dev-env.md:
```diff
@@ -0,0 +1,194 @@
+
+### Node Addresses
+
+Each AWS instance will receive both a private IPv4 address and a globally
```
So we're expecting to use DHCPv6?  Is that in line with what IPv6-using customers want?

russellb replied:
You're asking a good question about what customers may want.  In the context of the AWS dev/test env though, DHCPv6 is what we have available.

russellb commented:
This enhancement is no longer relevant.  We've solved the dev environment use case on bare metal for the most part.  We also started pursuing Azure as the first cloud env where we will enable testing, but that will come later.  /close
